### PR TITLE
Bug/4908 draft pools policy changes

### LINKED ISSUE #4908: Draft pools should not be accessible to non-admins
Only admins should be able to access non-published Pools/Advertisements.

Proposed solution:
Treat the Pool object as restricted to admins. Only expose the PoolAdvertisment object to the public, and ensure that PoolAdvertsiement queries do not return draft pools.
- Add the `@can(ability: "view")` directive to any query which returns a single Pool.
- Add the `@can(ability: "viewAdvertisement")` directive to any query which returns a single PoolAdvertisment.
- Add a `viewAdvertisment` method to the PoolPolicy class, which checks that the pool is published, and allows for null users (see DepartmentPolicy for how to allow null users).
- Add the `@can(ability: "viewAny")` directive to any query which returns multiple Pools.
- Add a scope to any query which returns an array of PoolAdvertisments, to scope by published.
- In the frontend, replace any Pool queries with PoolAdvertisement queries if possible (should cover all queries outside the Admin dashboard)

Acceptance Criteria
- [ ] Unless I am logged in as an Admin user, it should be impossible for me to infer the existence of a non-published pool via the api
- [ ] Unless I am logged in as an Admin user, it should be impossible for me to get any details about a non-published pool, even if I know the ID.

### PR DESCRIPTION
**Pull request by vd1992**
closes #4908 
makes changes to policies in schema and some frontend tweaks

### **TESTING**
will be important to test real cautiously to avoid introducing bugs or other problems
issue estimate is higher to reflect this

### ADMIN
login as an admin
go to /admin and bounce around (focus on pool related operations)
nothing should have changed
go to /talentsearch and repeat
nothing should have changed

### APPLICANT 
login as an applicant
go to /talentsearch
nothing should have changed from the stuff you should be able to do
get the id of a draft pool and attempt to navigate to `localhost:8000/en/browse/pools/:id`
you should get the graphQL oh no! message, but you will be able to view for PUBLISHED and EXPIRED

### CODE REVIEW

**Comment by vd1992** [comment]
few questions to resolve first, throwing back to in-progress
need to resolve what policy to put on `Pools`

File: frontend/talentsearch/src/js/components/search/SearchContainer.tsx [hunk 1]
```diff
@@ -339,10 +339,12 @@ const SearchContainerApi = () => {
...
   useEffect(() => {
-    if (searchFormData?.pools) {
-      setApplicantFilter({ pools: searchFormData?.pools });
+    if (searchFormData?.publishedPoolAdvertisements) {
+      setApplicantFilter({
+        pools: searchFormData?.publishedPoolAdvertisements,
+      });
     }
-  }, [searchFormData?.pools]);
+  }, [searchFormData?.publishedPoolAdvertisements]);
```
Comment (vd1992) [inline frontend/talentsearch/src/js/components/search/SearchContainer.tsx:347]: this resolves draft pools appearing on /search
it also prevents `countPoolCandidatesByPool` query from looking at draft pools

**Review by tristan-orourke — CHANGES_REQUESTED** [review]
I've left one suggested change for the schema.

Also, I see that the PoolCandidate object in the schema has both a `pool` and a `poolAdvertisment` attribute. This means that anyone with a PoolCandidate object has access to that Pool, even if they shouldn't according to the PoolPolicy. I think we can remove `pool` entirely from PoolCandidate, and replace any frontend operations that use it with `poolAdvertisement` without any trouble.

**Comment by vd1992** [comment]
@tristan-orourke 
no I think that would definitely be troublesome as the Snapshot functions with `Pool` inside `PoolCandidate`
changing to PoolAdvertisement would break Snapshots

image

**Comment by vd1992** [comment]
Snapshot creation comes from

image

snapshot test comes from

image

**Comment by tristan-orourke** [comment]
Ah, good point, Vachan. We'll leave the pool on PoolCandidate for now.

File: api/graphql/schema.graphql [hunk 2]
```diff
@@ -809,13 +812,17 @@ type Query {
...
+  pool(id: ID! @eq): Pool @find @can(ability: "view", query: true)
+  poolAdvertisement(id: ID! @eq): PoolAdvertisement
+    @find(model: "Pool")
+    @can(ability: "viewAdvertisement", model: "Pool", query: true)
+  poolAdvertisements: [PoolAdvertisement]!
+    @all(model: "Pool")
+    @can(ability: "viewAny", model: "Pool")
   publishedPoolAdvertisements: [PoolAdvertisement!]!
     @all(model: "Pool", scopes: ["wasPublished"])
   poolByKey(key: String! @eq): Pool @find
```
[the remark this answers is not in the document]
Comment (vd1992) [reply api/graphql/schema.graphql]: https://github.com/GCTC-NTGC/gc-digital-talent/pull/4915/commits/3e9c8be603c8b633a9719a8a56dd1dbc1c61c88c

**Review by tristan-orourke — CHANGES_REQUESTED** [review]
Awww, I found an issue :(
If I log in as an applicant and try to view a draft pool, I get the graphql error.
However, if I logout and try to view the draft pool as a guest... it works 😆

**Comment by petertgiles** [comment]
It did prevent me from viewing the pool, but the policy is not working well.  Looks like a null dereferencing problem.
image

**Comment by vd1992** [comment]
> It did prevent me from viewing the pool, but the policy is not working well. Looks like a null dereferencing problem. image

ah, it would probably be better to check if `$user` exists before calling methods on it
don't think it changes anything but here it is https://github.com/GCTC-NTGC/gc-digital-talent/pull/4915/commits/1c48bf013dde8965a22a211f32a874396e19ab30

**Comment by tristan-orourke** [comment]
Yep, I can't reproduce either. It's good!

**Review by tristan-orourke — APPROVED** [review]

**Pull request merged**